tmt/base.py

```diff
@@ -2341,7 +2341,7 @@ def import_plan(self) -> Optional['Plan']:
         self._imported_plan._original_plan = self
 
         # Update imported plan environment with the local environment
-        self._imported_plan.environment.update(self.environment)
+        self._imported_plan._environment.update(self.environment)
```

Comment:
Or, perhaps, we could set up a property setter instead to prevent similar surprises in the future?

Reply:
I am not sure. Here's an example of the same things being done in this snippet, and if you run them, you find that it works as intended:

```python
class A:
	def __init__(self):
		self.dict_obj={"value": "something"}
	
	@property
	def dict(self):
		return self.dict_obj

class B:
	def __init__(self):
		self.a = A()

	def update(self, other_dict):
		self.a.dict.update(other_dict)

b = B()
b.update({"other_value": "other"})
print(b.a.dict)
```

I am starting to think that there's more of an issue in the workflow that this segment is not called.

We should first get the test to cover the original issue here and trace that.

Reply:
> I am not sure. Here's an example of the same things being done in this snippet, and if you run them, you find that it works as intended:

The problem is that things are not the same :) In your code, `B.update()` touches `A.dict` which returns `A.dict_obj`, and `A.dict_obj` *does not change*, it's still the same object. On the other hand, check how `Plan.environment` starts:

```python
    @property
    def environment(self) -> EnvironmentType:
        """ Return combined environment from plan data and command line """
        if self.my_run:
            combined = self._plan_environment.copy()
```

Every time it's called (as long as `Plan.my_run` is set, of course), it starts with a fresh copy of `self._environment`, and therefore returns a brand new `dict` instance. Any `update()` on it will update just that fresh `dict`, no changes to it would be visible to any future calls of `Plan.environment`.

I'd try making `Plan.environment` a cached property:

* environment would be rendered once and cached,
* if we ever need to drop the rendered, cached environment, which we do need to, there's an official way of "clearing the cache`, `del plan.environment` - this move drops the value from `plan.__dict__`, and the first subsequent access to `plan.environment` would call the getter to produce a new value. This would make the "this change will affect the set of plan's envvars" moves more visible.
* can the imported plan pull in the environment from the importing plan? See the added lines in the example below - maybe there's a trap, but this would involve the importing plan without `Plan.import_plan()` explicitly updating neither `environment` nor `_environment`. And it might be just enough to fix the problem without even wandering into the `cached_property` territory.

```python
    @tmt.utils.cached_property
    def environment(self) -> EnvironmentType:
        if self.my_run:
            combined = self._plan_environment.copy()
            combined.update(self._environment)
            ...

            # We don't do this, but can we? It would make importing plan's environment
            # honored automagically.
            if self._original_plan is not None:
                combined.update(self._original_plan.environment)

            ...
            return combined
        return self._environment

    def _source_plan_environment_file(self) -> None:
        """ Add variables from the plan environment file to the environment """
        if (self.my_run and self.plan_environment_file.exists()
                and self.plan_environment_file.stat().st_size > 0):
            # Use __wrapped__ to force the reload of the file
            self._plan_environment = tmt.utils.environment_file_to_dict.__wrapped__(
                filename=self.plan_environment_file.name,
                root=self.plan_environment_file.parent,
                logger=self._logger)
            del self.environment
```

Not perfect, we still have to be aware of places that affect the environment, it would be much better for `self.environment` to manage the caching and changes, but that would be a way bigger change :/ We can do that one later (I also have an ugly WiP patch for making `environment` a bit smarter class, maybe I could implement the caching as well).

Reply:
> Every time it's called (as long as `Plan.my_run` is set, of course), it starts with a fresh copy of `self._environment`, and therefore returns a brand new `dict` instance. Any `update()` on it will update just that fresh `dict`, no changes to it would be visible to any future calls of `Plan.environment`.

Thanks for pointing it out. Indeed it seems like it would be trickier to fix this.

> can the imported plan pull in the environment from the importing plan?

I am in favor of this, especially since we might want to implement some of the points in #2340 to alter `context`, `adjust`, etc. as well. Not sure if it should be implemented in each property, but maybe there is some overlap with your ideas of updating `EnvironmentType` class?

Reply:
Pulling the environment from the original plan sounds like an interesting approach. That could work, and could make the things a bit more clear/clean.

Reply:
thanks for the discussion folks, so that is what I am doing.

Reply:
still in works sorry, I am away in the morning, but will fix this today

Reply:
updated, added tests, waiting for tests